Refactor PianoRoll

Veratil opened the pull request:
Kind of a bigger PR, but fixes a few things: #4098, #5008 (partial, but easy to add to editors now).

In addition to those fixes, it reworks how the piano roll was painted before (and makes it somewhat easier to follow). Local tests of ~1000 `paintEvent` frames using the demo `Skiessi-C64.mmpz` put `master`'s `paintEvent` at ~6-6.1ms average time, this PR runs at ~4-4.1ms (improvement of ~33%) on `DEBUG` build.

General summary of updates:
* Split out SongEditor's `positionLine` class into an updated separate class file `PositionLine`
* No longer can the key area be resized larger than `NumKeys * KEY_LINE_HEIGHT` (108 * 12) pixels
* If window is sized larger than max key area size, note edit area is resized instead
* Key area snaps to `KEY_LINE_HEIGHT` to prevent partial key draws (except if top key is black key, the previous key is drawn in that case)
* Position marker added to timeline
* Simplifies some math, some if branches, and draw code (old code went through white keys then through black keys, now only one for loop)

EDIT: I've updated the title to be a little more general since I'm touching a little more than just `painteEvent` with this PR. It still focuses on `paintEvent`, but there's a lot more to it at this point.

Veratil committed:
Rework paintEvent + side-effects

Veratil committed:
Finish final paintEvent rewrite, split out PositionLine class to own file

PhysSong commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2856,208 +2869,256 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
-			// draw a big one while checking if it is pressed or not
-			if( hasValidPattern() && m_pattern->instrumentTrack()->pianoModel()->isKeyPressed( key ) )
+			p.drawLine(x, keyAreaTop(), x, noteEditBottom());
+		}
```
It seems like you've added vertical lines to the note edit area. However, the color is inconsistent with that of the key area. Can't you extend the vertical line for the key area instead if you're going to keep those lines drawn?

Veratil replied:
Ah! I see what you mean. Minor oversight. Will fix that up. 👍

PhysSong reviewed (commented):
Here's my initial review for this PR.

Veratil commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2244,14 +2253,21 @@ void PianoRoll::mouseMoveEvent( QMouseEvent * me )
+			return;
+		}
+		int newHeight = height() - me->y();
+		if (height() - newHeight < KEY_AREA_MIN_HEIGHT)
```
Yes... yes it would be. :joy:

Veratil commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -1563,7 +1566,8 @@ void PianoRoll::mousePressEvent(QMouseEvent * me )
 	}
 
-	if( me->y() > keyAreaBottom() && me->y() < noteEditTop() )
+	if(me->button() == Qt::LeftButton &&
```
I figured this is "intended" behavior, but we can always revert it if necessary.

Veratil commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2856,208 +2869,256 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
-			}
-			// update y-pos
-			y -= WHITE_KEY_SMALL_HEIGHT;
+			// always increase m_notesEditHeight to compensate
```
Future enhancement. I'm currently reworking the mouse events. Once that is done I'm going to work on moving things to where they should be. I'd rather rework first to make sure behavior stays the same, then start moving things to where they should be handled.

JohannesLorenz replied:
Note: This will not be solved in this PR. I'll still leave it open as an information.

Veratil committed:
Fix lines and shade not drawing past the key area

Veratil committed:
Minor math fixup, remove unneeded subtraction

JohannesLorenz commented:
@Veratil This is marked WIP, but I think it's waiting for code review? Do you plan more enhancements/fixes in this PR currently?

Veratil commented:
> 
> 
> @Veratil This is marked WIP, but I think it's waiting for code review? Do you plan more enhancements/fixes in this PR currently?

Was waiting for other review/testing. It just needs to be rebased with recent updates it seems.

JohannesLorenz commented:
OK, removed "in progress".

Whoever reads this, please consider doing a review or testing the PR.

JohannesLorenz commented:
I'll do the review now (as we concluded that a review is still missing).

JohannesLorenz commented:
There's a problem with the PR base. You say you solve #4098, which is a 1.2.2 issue, but you want to merge into master, which will not fix that issue in 1.2.2.

So you need to decide:

1. Either rebase the whole PR on stable-1.2, or
2. Leave it on master and do not claim to solve 4098, or
3. Split it and only base the 4098 fix on stable-1.2 

I'd say use 3 if possible (It's a large amount of changes for a feature-freeze branch currently), if not use 1.

JohannesLorenz commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2856,208 +2869,256 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
+		int grid_line_y = keyAreaTop() + KEY_LINE_HEIGHT - 1;
+		static const int yTextCorrections[KeysPerOctave] = {
+		/*   C      D      E   F      G      A      B   */
+			-4, 0, -8, 0, -6, -4, 0, -8, 0, -8, 0, -6
```
This looks right with 4-space-tabs, but in case someone uses invalid tabs, I'd rather change this:

```suggestion
		    -4, 0, -8, 0, -6, -4, 0, -8, 0, -8, 0, -6
```

JohannesLorenz reviewed (dismissed):
Style/Code/Test Review finished, comments added.

PhysSong commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2856,208 +2869,256 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
+		pianoArea = keyAreaBottom() - keyAreaTop();
+		m_pianoKeysVisible = pianoArea / KEY_LINE_HEIGHT;
+		partialKeyVisible = pianoArea % KEY_LINE_HEIGHT;
+		if (m_pianoKeysVisible >= NumKeys)
```
It tests if the key area is tall enough to draw all keys.

JohannesLorenz replied:
OK, now I think it's clear, indeed. I'll resolve this comment.

Veratil commented:
> 
> 
> There's a problem with the PR base. You say you solve #4098, which is a 1.2.2 issue, but you want to merge into master, which will not fix that issue in 1.2.2.
> 
> So you need to decide:
> 
>     1. Either rebase the whole PR on stable-1.2, or
> 
>     2. Leave it on master and do not claim to solve 4098, or
> 
>     3. Split it and only base the 4098 fix on stable-1.2
> 
> 
> I'd say use 3 if possible (It's a large amount of changes for a feature-freeze branch currently), if not use 1.

Or 4. Say it's solved for master, and wontfix on 1.2.2 unless someone else wants to cherry pick and resolve all the conflicts thereabout.

JohannesLorenz commented:
> Or 4. Say it's solved for master, and wontfix on 1.2.2 unless someone else wants to cherry pick and resolve all the conflicts thereabout.

This issue is a crash on stable-1.2 and thus *must* be fixed there in some way (which excludes 2 btw). If you do (4), someone needs to cherry pick it to stable-1.2, and I don't know anyone who could cherry-pick it better then the author...

Veratil commented:
> 
> 
> > Or 4. Say it's solved for master, and wontfix on 1.2.2 unless someone else wants to cherry pick and resolve all the conflicts thereabout.
> 
> This issue is a crash on stable-1.2 and thus _must_ be fixed there in some way (which excludes 2 btw). If you do (4), someone needs to cherry pick it to stable-1.2, and I don't know anyone who could cherry-pick it better then the author...

That's understandable. There's two ways I can see to fix `stable-1.2`: 1) set the max height of the window (headers + pianoroll max + note edit min); or 2) stop drawing keys like this PR does.

For the max height idea, I was attempting to figure out how to set the max height first (but as mentioned this restricts the note edit window height at some point), but my efforts led to just the "mainWindow" of the piano roll being maxed. The actual window could be resized while the height of the `PianoRollWindow` would stop. It was very weird to me. I decided to work on rewriting the paintEvent instead. :)

As mentioned before, I would still say this PR fixes the issue though since iirc it happened on master as well hence me looking into this entire thing, so it is acceptable to mark it as such IMO.

JohannesLorenz commented:
> As mentioned before, I would still say this PR fixes the issue though since iirc it happened on master as well hence me looking into this entire thing, so it is acceptable to mark it as such IMO.

If you want to do this I'll create a duplicate of the issue for stable-1.2, which can then be fixed by a e.g. a cherry-picked PR.

JohannesLorenz commented:
@Veratil Do you plan to finish this PR? It's almost done...

Veratil committed:
Remove now unused m_oldNotesEditHeight

Veratil committed:
Resolve PR review suggestions

Veratil commented on include/PianoRoll.h:
```diff
@@ -40,6 +40,7 @@
 #include "ToolTip.h"
 #include "StepRecorder.h"
 #include "StepRecorderWidget.h"
+#include "PositionLine.h"
```
The latest commit tries to use forward declaration, seems it failed to build though...

JohannesLorenz replied:
It should have failed on your local computer, too.
You can fix it by adding the include into the affected cpp files.

Veratil replied:
I think it succeeded since I did an incremental build initially and it didn't complain. It fails now. 🤷‍♂️

JohannesLorenz commented:
Code and style are OK.

What remains is that someone needs to check whether it fixes #4098 - My monitor is not large enough.

PhysSong commented:
@JohannesLorenz You can move down the piano-roll window and resize it from its upper side. When you repeat that enough times, you'll get a large enough window.

JohannesLorenz commented:
Thanks @PhysSong that made it.

There's one issue still, though I'm not sure if we need to fix it: Open a piano roll, resize piano roll so it's getting big. Move it down to the bottom and resize it to the upper limit (that way, you enlarge the PianoRoll, thanks to @PhysSong ). Do that 3 or 4 times. Play a few notes. The lower notes are sometimes green covered, and the PianoRoll marks the wrong keys when I play keys. (It's not the same bug that I report for the stable PR)

JohannesLorenz commented:
I'm 100% happy with the PR. Testing finished.

Can I squash-and-merge it?

PhysSong commented:
> Can I squash-and-merge it?

@JohannesLorenz I think not now. #5442 may affect this PR.

Veratil commented:
> 
> 
> > Can I squash-and-merge it?
> 
> @JohannesLorenz I think not now. #5442 may affect this PR.

We need to decide which order to merge things, that's all.

PhysSong commented:
@Veratil Now it's your turn to fix conflicts with #5442. :smile:

Veratil commented:
> 
> 
> @Veratil Now it's your turn to fix conflicts with #5442. 😄

Rebased and adjusted code. 👍 Did some basic tests and zooms work still. Need a tester or two now!

PhysSong commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2873,214 +2878,257 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
+		p.setClipRect(0, keyAreaTop(), width(), keyAreaBottom() - keyAreaTop());
+		int key_line_y = keyAreaTop();
+		int grid_line_y = keyAreaTop() + m_keyLineHeight - 1;
+		static const int yTextCorrections[KeysPerOctave] = {
+		/*   C      D      E   F      G      A      B   */
+		    -4, 0, -8, 0, -6, -4, 0, -8, 0, -8, 0, -6
+		};
```
You can't use these constants anymore because of #5442.

Veratil replied:
I switched to using the one from the PR, and all the labels are way off now. 😩

Veratil replied:
I've managed to fix the offsets by doing a little different calculations, but with Y zoom I've found there are some draw bugs now when certain keys are the top key.

PhysSong commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2873,214 +2878,257 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
+			const int keyCode = key % KeysPerOctave;
+			const bool pressed = m_pattern->instrumentTrack()->pianoModel()->isKeyPressed(key);
+			const int yb = yt + keyHeight(key);
+			int kw, kh;
```
The compiler warns they may be used uninitialized, while it's not. Those warnings will make CI builds to fail.

Veratil replied:
Will be fixed in next commit.

Veratil committed:
Refactor PianoRoll Q_PROPERTYs

* Reduce code by using Q_PROPERTY's MEMBER function and removing getter/setter functions
  After looking at the getters and setters, they did nothing different than what direct
  access would allow. Nothing outside of PianoRoll used the public functions as well.
  Considering these factors we can reduce the number of functions by 2x the number of
  Q_PROPERTIES, and go with direct access instead.
* Remove need for keyboard pixmaps
  With the recent change to allow zooming vertically, aligning pixmaps is a PITA. Since
  we have themes which can take brushes and colors, it would be simpler to take a solid
  color or a gradient with some extra style properties to resize the keys and text colors.
  While it will slightly be a downgrade from pixmaps since they can be anything really,
  this will allow us to customize the piano roll further moving forward.
* Added the ability to update margins for TimeLineWidget and StepRecorderWidget
  These take a X coordinate, which was hardcoded to WHITE_KEY_WIDTH, and never looked
  back. Now we can adjust on the fly if we need to. Currently this just allows us to
  shift the left margin to the style-defined white key width.
* Some extra refactoring in paintEvent
  KEY_AREA_MIN_HEIGHT is now respected in case it tries to slide up too high. The
  constant has also been updated to be a multiple of DEFAULT_KEY_LINE_HEIGHT.

Veratil commented on data/themes/default/style.css:
```diff
@@ -165,6 +165,17 @@ PianoRoll {
 	qproperty-ghostNoteBorders: false;
 	qproperty-barColor: #078f3a;
 	qproperty-markedSemitoneColor: rgba(255, 255, 255, 30);
+	/* Piano keys */
```
I would like to note these updates are purely for testing and by no means are to be considered final.

Veratil committed:
Fix phantom pixmaps when PianoRoll not focused

Veratil committed:
Rebase to fix conflicts and fix note area resize bug

PhysSong commented:
I found a new bug.
1. Set the vertical zoom level to 50% and resize the piano roll large enough to show the whole key area
2. Change the vertical zoom level to 25%
3. The key area will disappear, and the resize bar is at a wrong place

Veratil commented:
> 
> 
> I found a new bug.
> 
>     1. Set the vertical zoom level to 50% and resize the piano roll large enough to show the whole key area
> 
>     2. Change the vertical zoom level to 25%
> 
>     3. The key area will disappear, and the resize bar is at a wrong place

Interesting. The resize bar is in the correct place actually, but the keys didn't get drawn for some reason. If you drag the resize bar up one pixel everything appears properly. I'll look into why they don't get drawn though.

Veratil committed:
Address vertical zoom bug

When showing all keys, and changing Y zoom down a level, the keys
would fail to draw. This was from a bounds error with m_startKey
being set to -108.

Veratil commented:
> 
> 
> > I found a new bug.
> > ```
> > 1. Set the vertical zoom level to 50% and resize the piano roll large enough to show the whole key area
> > 
> > 2. Change the vertical zoom level to 25%
> > 
> > 3. The key area will disappear, and the resize bar is at a wrong place
> > ```
> 
> Interesting. The resize bar is in the correct place actually, but the keys didn't get drawn for some reason. If you drag the resize bar up one pixel everything appears properly. I'll look into why they don't get drawn though.

Added some `printf`'s to see what was causing the issue. It looks like `m_startKey` is getting set to `-108`. Added some bounds checking to `m_startKey` and `m_totalKeysToScroll` so they would always be minimum 0.

PhysSong commented:
This one conflicts with #5543. @Veratil Could you resolve conflicts so we can test this one more time?

Veratil committed:
Update PositionLine class changes related to #5543

Veratil commented:
> 
> 
> This one conflicts with #5543. @Veratil Could you resolve conflicts so we can test this one more time?

Updated!

PhysSong commented:
If I select `classic` theme, I can't see keys.

Veratil commented:
> 
> 
> If I select `classic` theme, I can't see keys.

Oops, that's because I forgot to add the same properties to `classic`.

Spekular commented:
Just want to note that after this is merged, #669 should be easier to implement.

Veratil committed:
Add new Q_PROPERTYs to classic theme

Veratil commented:
@Spekular @PhysSong do we feel this is safe to merge yet? I want to do more, but I don't want to add to the huge changes for this already.

Spekular commented:
I haven't done a review so I'm not sure my opinion bears much weight (nothing stood out as problematic on a quick skim FWIW). What I can say is that I'm not opposed to the goals of the PR as described in the OP, all the changes mentioned sound good. Aside from that I'd like to defer to Johannes and PhysSong's judgement :P

JohannesLorenz commented:
I haven't watched it since April, so I can't say whether it's OK.

For the question about using a new PR, I'm always for making multiple small PRs - as you say, this one already got large enough. :smile:

PhysSong commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2800,353 +2655,337 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
-		case PR_WHITE_KEY_SMALL:
-			if( prKeyOrder[( ( m_startKey + 1 ) %
-					KeysPerOctave)] != PR_BLACK_KEY )
+		int pianoArea, partialKeyVisible, topKey, topNote;
+		pianoArea = keyAreaBottom() - keyAreaTop();
+		m_pianoKeysVisible = pianoArea / m_keyLineHeight;
+		partialKeyVisible = pianoArea % m_keyLineHeight;
```
Very minor, but `pianoAreaHeight` looks clearer.

PhysSong commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -3389,70 +3230,92 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
+		SCROLLBAR_SIZE,
+		height() - PR_TOP_MARGIN - SCROLLBAR_SIZE
+	);
+	int pianoArea = keyAreaBottom() - PR_TOP_MARGIN;
+	int numKeysVisible = pianoArea / m_keyLineHeight;
```
The same.

PhysSong commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -2800,353 +2655,337 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
 			{
-				p.setPen( lineColor() );
+				// draw next white key
+				drawKey(key - 1, grid_line_y + m_keyLineHeight);
+				drawHorizontalLine(key - 1, grid_line_y + m_keyLineHeight);
```
If `lastKey + 1` is a white key, the loop will stop after drawing it, whereas it will draw `lastKey` if `lastKey + 1` is a black key. Is this the intended behavior?

Veratil replied:
This handles the case where we draw a black key over a white key (current behavior). So, yes it's intended for that purpose.

PhysSong commented:
Also, you changed the width of black keys to the same as white keys. Although [it's themeable](https://github.com/LMMS/lmms/issues/3751#issuecomment-623147961), don't we need discussions when changing the default?

Veratil committed:
Reset black piano keys to be shorter; add some comments

Veratil commented:
> 
> 
> Also, you changed the width of black keys to the same as white keys. Although [it's themeable](https://github.com/LMMS/lmms/issues/3751#issuecomment-623147961), don't we need discussions when changing the default?

Accidental leftover while testing. Changed down to 48 pixels.

PhysSong commented:
There might be something I missed, but I think it looks okay now.

Veratil commented:
If no one opposes, I'll merge this in 2 days.
